wallet/src/store.rs

```diff
@@ -0,0 +1,1196 @@
+    }
+
+    fn get_transactions(
+        &self,
+        lock_hash: &H256,
+        skip_num: usize,
+        take_num: usize,
+    ) -> Vec<CellTransaction> {
+        let iter = self
+            .db
+            .iter(COLUMN_LOCK_HASH_TRANSACTION, lock_hash.as_bytes())
+            .expect("wallet db iter should be ok");
+        iter.skip(skip_num)
+            .take(take_num)
+            .take_while(|(key, _)| key.starts_with(lock_hash.as_bytes()))
+            .map(|(key, value)| {
+                let consumed_by: Option<TransactionPoint> =
+                    deserialize(&value).expect("deserialize TransactionPoint should be ok");
+                let lock_hash_index = LockHashIndex::from_slice(&key);
+                CellTransaction {
+                    created_by: lock_hash_index.into(),
+                    consumed_by,
+                }
+            })
+            .collect()
+    }
+
+    fn get_lock_hash_index_states(&self) -> HashMap<H256, LockHashIndexState> {
+        self.db
+            .iter(COLUMN_LOCK_HASH_INDEX_STATE, &[])
+            .expect("wallet db iter should be ok")
+            .map(|(key, value)| {
+                (
+                    H256::from_slice(&key).expect("db safe access"),
+                    deserialize(&value).expect("deserialize LockHashIndexState should be ok"),
+                )
+            })
+            .collect()
+    }
+
+    fn insert_lock_hash(
+        &self,
+        lock_hash: &H256,
+        index_from: Option<BlockNumber>,
+    ) -> LockHashIndexState {
+        // need to lock chain state, avoids inconsistent state in processing
+        let chain_state = self.shared.lock_chain_state();
+        let index_state = LockHashIndexState {
+            block_number: chain_state.tip_number(),
+            block_hash: chain_state.tip_hash().to_owned(),
+        };
+        self.commit_batch(|batch| {
+            if let Some(from_block_number) = index_from {
+                let mut index_lock_hashes = HashSet::new();
+                index_lock_hashes.insert(lock_hash.to_owned());
+                (from_block_number..=chain_state.tip_number()).for_each(|block_number| {
+                    let block = self
+                        .shared
+                        .block_hash(block_number)
+                        .and_then(|hash| self.shared.block(&hash))
+                        .expect("block exists");
+                    self.attach_block(batch, &index_lock_hashes, &block);
+                });
+            }
+            batch.insert_lock_hash_index_state(lock_hash, &index_state);
+        });
+        index_state
+    }
+
+    fn remove_lock_hash(&self, lock_hash: &H256) {
+        self.commit_batch(|batch| {
+            let iter = self
+                .db
+                .iter(COLUMN_LOCK_HASH_LIVE_CELL, lock_hash.as_bytes())
+                .expect("wallet db iter should be ok");
+
+            iter.take_while(|(key, _)| key.starts_with(lock_hash.as_bytes()))
```

Comment:
we can't construct an `end_key` here for `delete_range`

Reply:
All the keys you want to delete here starts with `lock_hash`, you can just treat `lock_hash` as an integer, add 1 to it, you will get the first key succeeding all the keys you need to delete here.

Reply:
we need to delete `cell_out_point_lock_hash` which is a reverse index of lock hash, it can't utilize the `delete_range` function, to make it simple, I keep the same `iter/delete` pattern in `delete_lock_hash_transaction`.